art/estimators/object_tracking/pytorch_goturn.py

```diff
@@ -0,0 +1,626 @@
+            output_width = max(math.ceil(compute_output_width_f(bbox_tight)), roi_width)
+            output_height = max(math.ceil(compute_output_height_f(bbox_tight)), roi_height)
+            if image.ndim > 2:
+                # output_image = np.zeros((int(output_height), int(output_width), image.shape[2]), dtype=image.dtype)
+                output_image = torch.zeros((int(output_height), int(output_width), image.shape[2]), dtype=image.dtype)
+            else:
+                # output_image = np.zeros((int(output_height), int(output_width)), dtype=image.dtype)
+                output_image = torch.zeros((int(output_height), int(output_width)), dtype=image.dtype)
+
+            # edge_spacing_x = min(bbox_tight.edge_spacing_x(), (image.shape[1] - 1))
+            # edge_spacing_y = min(bbox_tight.edge_spacing_y(), (image.shape[0] - 1))
+            edge_spacing_x = min(edge_spacing_x_f(bbox_tight), (image.shape[1] - 1))
+            edge_spacing_y = min(edge_spacing_y_f(bbox_tight), (image.shape[0] - 1))
+
+            # rounding should be done to match the width and height
+            output_image[
+                int(edge_spacing_y) : int(edge_spacing_y) + cropped_image.shape[0],
+                int(edge_spacing_x) : int(edge_spacing_x) + cropped_image.shape[1],
+            ] = cropped_image
+
+            return output_image, pad_image_location, edge_spacing_x, edge_spacing_y
+
+        target_pad, _, _, _ = crop_pad_image(prev_bbox, prev_frame)
+        cur_search_region, search_location, edge_spacing_x, edge_spacing_y = crop_pad_image(prev_bbox, curr_frame)
+
+        target_pad_in = self._preprocess(target_pad).unsqueeze(0).to(self._device)
+        cur_search_region_in = self._preprocess(cur_search_region).unsqueeze(0).to(self._device)
+
+        pred_bb = self._model.forward(target_pad_in.float(), cur_search_region_in.float())
+
+        pred_bb = torch.squeeze(pred_bb)
+
+        k_scale_factor = 10
+        height = cur_search_region.shape[0]
+        width = cur_search_region.shape[1]
+
+        # Normalize the image bounding box
+        pred_bb[0] = pred_bb[0] / k_scale_factor * width
+        pred_bb[2] = pred_bb[2] / k_scale_factor * width
+        pred_bb[1] = pred_bb[1] / k_scale_factor * height
+        pred_bb[3] = pred_bb[3] / k_scale_factor * height
+
+        # brings gradients to zero
+        # pred_bb = torch.round(pred_bb)
+
+        # move the bounding box to target/search region coordinates
+        raw_image = curr_frame
+        pred_bb[0] = max(0.0, pred_bb[0] + search_location[0] - edge_spacing_x)
+        pred_bb[1] = max(0.0, pred_bb[1] + search_location[1] - edge_spacing_y)
+        pred_bb[2] = min(raw_image.shape[1], pred_bb[2] + search_location[0] - edge_spacing_x)
+        pred_bb[3] = min(raw_image.shape[0], pred_bb[3] + search_location[1] - edge_spacing_y)
+
+        # brings gradients to zero
+        # pred_bb = torch.round(pred_bb)
```

Comment:
Please remove all comment lines if they are not needed.

Reply:
I agree and will remove them in the upcoming patch releases. I would like to keep them for now for documentation and future debugging.